apps/example-next/app/(react-query-streaming)/test-streaming/page.tsx

```diff
@@ -3,19 +3,33 @@ import { HydrationBoundary, dehydrate } from '@tanstack/react-query';
 import { tsr } from '../../react-query-utils/tsr';
 import { ClientComponentSuspense } from './client-component-suspense';
 import { getQueryClientRsc } from '../../react-query-utils/get-query-client';
+import type { Metadata } from 'next';
 
 export const dynamic = 'force-dynamic';
 
+const queryData = {
+  params: { id: 2 },
+  query: { foo: 'test', bar: 123 },
+};
+
+export async function generateMetadata(): Promise<Metadata> {
```

Comment:
It's better to have one example use-case per page. It will make it easier to find, and prevent any inadvertent confusion that the two use-cases are tied to each other, if someone is not well-versed enough in Next. In this case, someone might think that streaming and metadata are tied to each other.

We are also running the same query twice once in `generateMetadata()` and once in `Test()` but on two separate query clients, so the query will actually run twice, but the underlying `fetch` request will end up being deduplicated by Next.js pre-v15, and on v15+ it will actually run two fetches. This could be intended or it could be not, but it's better to make sure our example is unopinionated as there are different strategies of either using separate query clients, or using a cached query client per-request.

So it would be much better and avoid any potential confusion, if we have a separate example in something like `app/(react-query)/test-metadata/page.tsx` where we make sure we do not run the same query twice in the server component itself and the `generateMetadata()` function.

Reply:
https://nextjs.org/blog/next-15-rc#caching-updates

And in the official docs https://nextjs.org/docs/app/api-reference/functions/fetch#optionscache:~:text=Good%20to%20know,in%20Next.js.

There's two things at play here.

React itself is memoizing `fetch` requests in RSC across multiple `fetches` during the same request using `React.cache`. This is the behavior in React 18, and is going to be [changed in React 19](https://togithub.com/facebook/react/pull/28896) where React will not patch the `fetch` function at all. This kind of caching shouldn't cause problems anyway as it is only memoized per-request and is not persisted. You most likely really want to only call `fetch` once for any duplicated fetches on the same page. This is the problem that react-query itself has always fixed since it maintains its own cache, and does not refetch within `staleTime`. So in React 19 and/or Next.js 15, you will need to actually make sure you use a [single `QueryClient` instance](https://tanstack.com/query/latest/docs/framework/react/guides/advanced-ssr#alternative-use-a-single-queryclient-for-prefetching) across multiple server components and functions like `generateMetadata` to make sure you do not actually fetch multiple times. This is even mentioned in the RQ docs.

> Next.js already dedupes requests that utilize fetch(), but if you are using something else in your queryFn, or if you use a framework that does not dedupe these requests automatically, using a single queryClient as described above might make sense, despite the duplicated serialization.

Next.js 15 and React 19 have still not been fully released, so probably why the RQ docs have not been updated to reflect the change in this caching/memoization behavior.

The one probably causing you issues is the Next.js persistent data cache. Next.js will actually cache fetches across multiple requests and even deployments. This is explained here https://nextjs.org/docs/app/building-your-application/caching#data-cache

You can control this behavior using the [`fetchOptions.cache`](https://github.com/ts-rest/ts-rest/blob/main/apps/example-next/app/(react-query-streaming)/test-streaming/client-component-suspense.tsx#L12) option in ts-rest clients for single requests or change the default behavior for any fetches across a Next.js route segment using the [fetchCache](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#fetchcache) segment config